doc/man/man1/erun.1
.\" Manpage for erun
.TH man 8 "June 2026" "7.0" "erun man page"
.SH NAME
erun \- General EAR configuration information.
.SH SYNOPSIS
erun
.SH DESCRIPTION
erun is a program that simulates all the SLURM and EAR SLURM Plug-in pipeline. It was designed to provide compatibility between MPI implementations not fully compatible with SLURM SPANK plug-in mechanism (e.g., OpenMPI), which is used to set up EAR at job submission.

The syntax to run an MPI application with erun has the form 
.B mpirun -n <X> erun --program='my_app arg1 arg2 .. argN'.
With this, mpirun will run X erun processes. Then, erun will launch the application my_app with the arguments passed, if specified. One can use as many parameters as needed, but the semicolons have to cover all of them in case there are more than just the program name.

erun will simulate on the remote node both the local and remote pipelines for all created processes. It has an internal system to avoid repeating functions that are executed just one time per job or node, like SLURM does with its plugins.

.SH OPTIONS
.SS erun options

.RS
.nf
	--nodes=<arg>       Sets the number of nodes.
	--program=<arg>     Sets the program to run.
	--clean             Removes the internal files.
.RE

.SS SLURM options
.RS
.nf
	--ear               Enables/disables Energy Aware Runtime Library
	--ear-policy		
	--ear-cpufreq       Specifies the start frequency to be used by EAR policy (in KHz)
	--ear-policy-th     Specifies the threshold to be used by EAR policy (max 2 decimals) {value=[0..1]}
	--ear-user-db       Specifies the file to save the user applications metrics summary 'file.nodename.csv' file will be created per node. If not defined, these files won't be generated.
	--ear-verbose       Specifies the level of the verbosity{value=[0..1]}; default is 0
	--ear-learning      Enables the learning phase for a given P_STATE {value=[1..n]}
	--ear-tag           Sets an energy tag (max 32 chars)
.RE


.SH NOTES
To launch n applications with erun command through an sbatch job, one must set the environment variable SLURM_STEP_ID to values from 0 to n-1 before each mpirun call. This way, erun will inform the EARD the correct step ID to be stored then to the Database.

The --job-id and --nodes parameters create the environment variables that SLURM would have created automatically, because it is possible that your application make use of them. The --clean option removes the temporal files created to synchronize all ERUN processes.

One must load the EAR environment module or define its environment variables in the environment or script for erun to function properly.

.TS
tab(@), right, box;
c | c
rB | r.
Variable@Parameter
_
EAR_INSTALL_PATH=<path>@prefix=<path>
EAR_TMP=<path>@localstatedir=<path>
EAR_ETC=<path>@sysconfdir=<path>
EAR_DEFAULT=<on/off>@default=<on/off>
.TE

.SH BUGS
No known bugs.
.SH AUTHOR
EAR suport team (ear-support@bsc.es)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

### Changed
- New energy node API.

### Fixed

## [7.0.0] - 2026-06-19

### Added
- Secret key between EARDBD and EARDBD API.
- Added a HWMON function to close unused file descriptors.
- Added debug callback proposal.
- Added count system to HSMP.
- Improve overhead class to add a runtime selection system.
- Added SCOPE_NODE and SCOPE_JOB to different APIs.
- MIG slices study and adding a GPU topology function.
- Read the cluster conf in eardbd.
- Add an ENERGY (NODE) plugin to read the E4 energy (and potentially, other energy counters based on scripts).
- Add API DEMO to GPU.
- Power cap verification functions.
- Add citation file.

### Changed
- CPI, cache, temperature, flops, cpufreq, bandwidth, gpu and imcefreq APIs.
- Improved timestamp_diff() code robustness.
- Improve the locking and retry systems in sockets.
- Plugin manager 2.0.
- Database naming scheme revision.
- Update commands --help and man pages.

### Fixed

- Fixed plugin_manager_close() function.
- Fix SIGSEGV when the hostname prefix is too large.
- EARL-EARD timeouts.
- Wall Werror CFLAGS fixes.
- Prevent SIGSEGV when passing a NULL energy tag.
- CSV files mode and ownership.

## [6.3.1] - 2026-06-04

### Changed

- Disable data center monitor's `conf` plugin if there is some trouble when reading configuration.
- Updated v5.2.4's CHANGELOG to mention security issues were found by [REQON B.V.]((https://reqon.nl)).

### Fixed

- Fixed `--with-slurm` configure flag management.
- Node parsing error.
- Search for cmocka libs in lib64 too.
- Prevent checking for freeipmi if not requested at configure time.
- Fixed CPUPOW\_F\_POWERCAP\_GET(mgt\_cpupow\_amd17\_powercap\_get) function reps initialization.
- Econtrol fixes:
    - Ranges loop when sending message to whole cluster. There was an error coming from the extension to support ranges with ranges.
    - When using --host the len of the hostnames was truncated becuase \\0 character was not taken into account.

## [6.3.0] - 2026-03-16

### Added

- Automatic release done through Gitlab CI.

### Changed

- The rpm spec file installs the doc directory and module file.

### Fixed

- Resolve "Policies default PState is overwritten in EARD's main function".
- Resolve "Hardcoded ear user in ear\_create\_tmp\_folder".
- Fix the HWMON ARCH in ENERGY\_CPU when the NODE energy is PKG energy is mistaken for NODE energy.
- Resolve "erun  blocked when no having permission to access job directory".

## [6.2.0] - 2026-02-24

### Added
- Add m4 files for curl, jansson and microhhtpd.
- EAROwner added.

### Changed
- EARD security fixes.
- Improve file permission and ownership.
- `erun` command sets JobID to 1 by default.
- Default CFLAGS changed to add `-Wall -Werror -O2 -g` to improve code quality.
- Lighter CI pipeline.

### Fixed
- Prevent calling `mgt\_cpufreq\_governor\_available` when the symbol is NULL.
- Fix multiple opened fd checks that might be zero.
- Fixed setuid security issues in ereport, eacct, ear-info and econtrol.
- EARD rAPI: Generic message is not propagating properly.
- IPMI sensors.
- Fixed "Policies default PState is overwritten in EARD's main function."

## [6.1.0] - 2025-12-18

### Added
- Add the `EAR_ENERGY_ARG` envvar into `energy_node/node/ipmi_sensors.c`.

### Changed
- CI pipelines.
- pread limits.

### Fixed
- Serialize/Deserialize at dummy cluster\_conf.
- Resolve "Error found in HWMON class in ENERGY CPU".
- EARD local API timeouts.
- Resolve "Warning fixes to include -Wall -Werror to CI/CD".
- IPMI fixes.

## v6.0 - 2025-10-29

### Added
- v6.0 GPU GFLOPS field added in the EAR Database.
- v6.0 Added quality of life functions to management/cpupow API.
- v6.0 New API for reading metrics through Linux hwmon.
- v6.0 NVIDIA Grace Hopper and Grace Grace devices support.
- v6.0 Added AdminUsers to provide more privileges than AuthorizedUsers.
- v6.0 Added --force option to the erun command.
- v6.0 Added a mechanism to dynamically increase the timeout window in msg\_internals when a node connects but does not return data.
- v6.0 Added stall cycles to CPI API.
- v6.0 Application CPU usage field added in the EAR Database.
- v6.0 Temperature readings through Linux hwmon.
- v6.0 Added ARM compatibility to PERF (still unstable).
- v6.0 New tool to create application signature from loop signature in DB.
- v6.0 New command enodectl.
- v6.0 Added HSMP opening mode control.
- v6.0 Added an option to ereport to query by user\_group.
- v6.0 IPMI API
- v6.0 Non-interactive version of the edb\_create command.
- v6.0 Now the topology module stores a socket index.
- v6.0 Node sharing strategy added.
- v6.0 Added non-computational power support to soft powercap.

### Changed
- v6.0 Renamed DBIP and DBSECIP in ear.conf to EARDBD\_IP and EARDBD\_MIRROR\_IP to better reflect what it is.
- v6.0 Energy node API.
- v6.0 EAR Loader improved.
- v6.0 Bandwidth, cache and flops APIs improved.
- v6.0 Create a fallback mechanism for failed inserts to DB.
- v6.0 Added pointer recalculation in cache\_data\_copy.
- v6.0 Changed eacct's energy field from Joules to KWh.
- v6.0 eacct's -n flag now just counts for jobs.
- v6.0 EAR source code follows clang-format style.
- v6.0 GPU energy savings are reported if a GPU optimization policy is being used.

### Fixed
- v6.0 A heap-buffer overflow when opening perf files is fixed.
- v6.0 Minor fixes in the database size calculator tool.
- v6.0 Fixed erroneous time weight calculation for each loop in loops\_average.
- v6.0 Fixed SIGALRM for job validation.
- v6.0 Switch to P\_STATE 0 after userspace governor in AMD17 architecture, and a fix in PCI opened file descriptors.
- v6.0 Fixed DUMMY Cache API.
- v6.0 Fixed compilation errors when using PostgreSQL.

src/slurm_plugin/erun.c
/***************************************************************************
 * Copyright (c) 2024 Energy Aware Runtime - Barcelona Supercomputing Center
 *
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 **************************************************************************/

/* clang-format off */
#include <common/utils/args.h>
#include <slurm_plugin/erun_lock.h>
#include <slurm_plugin/erun_signals.h>
#include <slurm_plugin/slurm_plugin.h>
#include <slurm_plugin/slurm_plugin_environment.h>
#include <slurm_plugin/slurm_plugin_serialization.h>
#include <sys/wait.h>

//
extern plug_serialization_t sd;
// Buffers
char plug_tmp[SZ_NAME_LARGE];
char plug_etc[SZ_NAME_LARGE];
char plug_pfx[SZ_NAME_LARGE];
char plug_def[SZ_NAME_LARGE];
char path_app[SZ_PATH];
char path_tmp[SZ_PATH];
char path_sys[SZ_PATH];
char buffer[SZ_PATH];
//
static int _inactive;
static int _error;
static int _force;
static int _clean;
static int _help;
//
static int _at;
int _sp;
//
static char *args[128];
char **_argv;
int _argc;

int plug_is_action(int _ac, int action)
{
    return _ac == action;
}

int help(int argc, char *argv[])
{
    printf("Usage: %s [OPTIONS]\n", argv[0]);
    printf("\nOptions:\n");
    //	printf("\t--job-id=<arg>\t\tSet the JOB_ID.\n");
    printf("\t--nodes=<num>\t\tSets the number of nodes.\n");
    printf("\t--args <args>\t\tSets the program to run and its arguments.\n");
    printf("\t--program=<arg>\t\tSets the program to run. (Deprecated).\n");
    //	printf("\t--plugstack [ARGS]\tSet the SLURM's plugstack arguments. I.e:\n");
    //	printf("\t\t\t\t--plugstack prefix=/hpc/opt/ear default=on...\n");
    printf("\t--force\t\t\tForces the ERUN execution when has already passed by SRUN.\n");
    printf("\t\t\t\tIt doesn't connect EARD twice.\n");
    printf("\t--clean\t\t\tRemoves the internal files.\n");
    printf("SLURM options:\n");

    return 0;
}

int pipeline(int argc, char *argv[], int sp, int at)
{
    _sp = sp;
    _at = at;

    if (plug_is_action(_at, Action.init)) {
        slurm_spank_init(_sp, argc, argv);

        if (plug_context_is(_sp, Context.srun)) {
            slurm_spank_init_post_opt(_sp, argc, argv);
        } else if (plug_context_is(_sp, Context.remote)) {
            slurm_spank_user_init(_sp, argc, argv);
            // The master calls user_init before any task is calling
            // task_init, because all the other tasks waits the master
            // to finish its remote pipeline completely. Then there
            // isn't required a task wait here.
            slurm_spank_task_init(_sp, argc, argv);
        }
    } else if (plug_is_action(_at, Action.exit)) {
        if (plug_context_is(_sp, Context.remote)) {
            slurm_spank_task_exit(_sp, argc, argv);
            if (sd.erun.is_master) {
                slurm_spank_exit(_sp, argc, argv);
            }
        }
        if (plug_context_is(_sp, Context.srun)) {
            slurm_spank_exit(_sp, argc, argv);
        }
    }

    return ESPANK_SUCCESS;
}

int print_argv(int argc, char *argv[])
{
    int i;
    int j;

    for (i = 0, j = 0; i < argc; ++i) {
        strcpy(&buffer[j], argv[i]);
        j += strlen(argv[i]) + 1;
        buffer[j - 1] = ' ';
    }
    buffer[j] = '\0';
    plug_verbose(_sp, 4, "input: %s", buffer);
    return 0;
}

static void parse_program(char *program, char *args[])
{
    int length = strlen(program);
    int i; // Character iterator
    int j; // Argument iterator

    args[0] = program;
    for (i = 0, j = 1; i < length; ++i) {
        if (program[i] == ' ' || program[i] == '\n') {
            // If detected space, we split the string by setting EOF
            program[i] = '\0';
            // If next character is not EOF, space, new line or backlash,
            // we consider it as argument.
            if (program[i + 1] != '\0' && program[i + 1] != ' ' &&
                program[i + 1] != '\n' && program[i + 1] != '\\') {
                args[j] = &program[i + 1];
                ++j;
            }
        }
    }
    args[j] = NULL;
    plug_verbose(_sp, 4, "ERUN --program decomposition:");
    for (i = 0; args[i] != NULL; ++i) {
        plug_verbose(_sp, 4, "%d: '%s'", i, args[i]);
    }
}

static int parse_args(char *argv[], char *program, char *args[])
{
    int j = 0; // Argument iterator
    while (argv != NULL && *argv != NULL) {
        if (strncmp(*argv, "--args", 6) == 0) {
            break;
        }
        ++argv;
    }
    if (strncmp(*argv, "--args=", 7) == 0) {
        if (strlen(&(*argv)[7]) == 0) {
            return 0;
        }
        args[j++] = &(*argv)[7];
    }
    ++argv;
    while (argv != NULL && *argv != NULL) {
        args[j++] = *argv;
        ++argv;
    }
    return 0;
}

static char *getenv_pbs_jobid()
{
    static char job_id[32];
    char *c;
    char *p;
    if (!(p = getenv("PBS_JOBID"))) {
        return NULL;
    }
    if ((c = strchr(p, '[')) != NULL) {
        *c = '\0';
    }
    if ((c = strchr(p, '.')) != NULL) {
        *c = '\0';
    }
    sprintf(job_id, "%s", p);
    return job_id;
}

int job(int argc, char *argv[])
{
    int err_pfx = 1;
    int err_tmp = 1;
    int err_etc = 1;
    int err_def = 1;
    char *p     = NULL;
    int i       = 0;

    // Its ERUN
    sd.erun.is_erun = 1;
    // Setting ERUN reference
    setenv(Var.is_erun.ear, "1", 1);
    // Enabling plugin and ERUN components
    setenv(Var.comp_plug.comp, "1", 1);

    // Arguments processing
    if (args_get(argc, argv, "clean", NULL)) {
        _clean = 1;
    }
    if (args_get(argc, argv, "help", NULL)) {
        _help = 1;
    }
    if (args_get(argc, argv, "force", NULL)) {
        _force = 1;
    }
    if ((p = args_get(argc, argv, "program", NULL)) != NULL) {
        sprintf(path_app, "%s", p);
        parse_program(path_app, args);
        // This is safe because the parsing has already split the program string
        setenv(Var.name_app.slurm, path_app, 1);
    } else if ((p = args_get(argc, argv, "args", NULL)) != NULL) {
        parse_args(argv, p, args);
    } else {
        _help = !_clean;
    }
    // Converting configuration enrivonment variables
    // (INSTALL_PATH, ETC, TMP) in input parameters.
    _argc           = argc + 4;
    _argv           = malloc(sizeof(char **) * _argc);
    _argv[argc + 0] = plug_def;
    _argv[argc + 1] = plug_pfx;
    _argv[argc + 2] = plug_etc;
    _argv[argc + 3] = plug_tmp;

    for (i = 0; i < argc; ++i) {
        _argv[i] = argv[i];
    }
    if ((p = getenv("EAR_INSTALL_PATH")) != NULL) {
        sprintf(plug_pfx, "prefix=%s", p);
        err_pfx = 0;
    }
    if ((p = getenv("EAR_ETC")) != NULL) {
        sprintf(plug_etc, "sysconfdir=%s", p);
        err_etc = 0;
    }
    if ((p = getenv("EAR_TMP")) != NULL) {
        sprintf(plug_tmp, "localstatedir=%s", p);
        sprintf(path_tmp, "%s", p);
        err_tmp = 0;
    } else {
        sprintf(path_tmp, "/tmp");
    }
    if ((p = getenv("EAR_DEFAULT")) != NULL) {
        sprintf(plug_def, "default=%s", p);
        err_def = 0;
    } else {
        sprintf(plug_def, "default=on");
        err_def = 0;
    }
    for (i = 0; i < argc; ++i) {
        _argv[i] = argv[i];
    }

    // Getting N_NODES
    char *n_nodes;
    for (i = 0; i < argc; ++i) {
        if ((strlen(argv[i]) > 7) && (strncmp("--nodes=", argv[i], 8) == 0)) {
            setenv(Var.job_node_count.slurm, &argv[i][8], 1);
        }
    }
    // OpenMPI version of N_NODES
    if ((n_nodes = getenv(Var.job_node_count.slurm)) == NULL) {
        char *size_world = getenv("OMPI_COMM_WORLD_SIZE");
        char *size_local = getenv("OMPI_COMM_WORLD_LOCAL_SIZE");
        char buffer[8];

        if (size_world != NULL && size_local != NULL) {
            i = atoi(size_world) / atoi(size_local);
            sprintf(buffer, "%d", i);
            n_nodes = buffer;
        }
    }
    if (n_nodes != NULL) {
        setenv(Var.step_node_count.slurm, n_nodes, 1);
    }

    // Getting JOB_ID
    char *job_id;
    char *step_id;
    // Patching other job managers
    if (!(job_id = getenv(Var.job_id.slurm))) {
        if (!(job_id = getenv("OAR_JOB_ID")))
            if (!(job_id = getenv_pbs_jobid()))
                if (!(job_id))
                    job_id = "0";
        // Setting the SLURM version of the variable
        setenv(Var.job_id.slurm, job_id, 1);
    }
    if ((step_id = getenv(Var.step_id.slurm)) == NULL) {
        if (step_id == NULL)
            step_id = "0";
        // Setting the SLURM version of the variable
        setenv(Var.step_id.slurm, step_id, 1);
    } else {
        sd.erun.is_step_id = 1;
    }
    // Saving variables
    sd.erun.job_id  = atoi(job_id);
    sd.erun.step_id = atoi(step_id);

    // Input parameters final
    print_argv(_argc, _argv);

    // Going inactive?
    _inactive = isenv_agnostic(_sp, Var.ctx_was_srun.mod, "1");

    if (_inactive) {
        if (_force) {
            sd.erun.avoid_connecting = 1;
        } else {
            _error = 2;
        }
    }
    if (_inactive && !_force) {
        _error = 2;
    }
    if (err_pfx | err_etc | err_tmp | err_def) {
        _error = 1;
    }
    return 0;
}

int step(int argc, char *argv[])
{
    // If STEP_ID wasn't read
    if (!sd.erun.is_step_id) {
        sprintf(buffer, "%d", sd.erun.step_id);
        setenv("SLURM_STEP_ID", buffer, 1);
    }
    // Currently, SLURM_LOCALID is only used to determine the master node. In
    // case its value gains new uses, a lock would have to be used to numerate
    // each local task id.
    if (getenv(Var.local_id.slurm) == NULL) {
        if (sd.erun.is_master) {
            setenv(Var.local_id.slurm, "0", 1);
        } else {
            setenv(Var.local_id.slurm, "1", 1);
        }
    }
    plug_verbose(_sp, 2, "program: '%s'", path_app);
    plug_verbose(_sp, 2, "job/step id: '%d/%d'", sd.erun.job_id, sd.erun.step_id);

    return 0;
}

static int execute(int argc, char *argv[])
{
    pid_t fpid;
    //
    fpid = fork();

    if (fpid == 0) {
        // Setting SLURMs task pid
        sprintf(buffer, "%d", getpid());
        setenv(Var.task_pid.ear, buffer, 1);

        // Executting
        if (execvp(args[0], args) == -1) {
            plug_verbose(_sp, 0, "failed to run the program (%s: %s)", args[0], strerror(errno));
            exit(0);
        }
        // No return
    } else if (fpid == -1) {
        plug_verbose(_sp, 0, "failed to run the program (fork)");
    } else {
        waitpid(fpid, &sd.subject.exit_status, 0);
        plug_verbose(_sp, 2, "program returned with status %d", sd.subject.exit_status);
    }

    return 0;
}

int main(int argc, char *argv[])
{
    // Creating job and reading arguments
    job(argc, argv);

    // Signals initialization
    signals();

    // Clean
    if (_clean) {
        // Cleaning $EAR_TMP/erun folder
        plug_verbose(_sp, 0, "cleaning erun folder");
        all_clean(path_tmp);
        return 0;
    }

    // Help
    if (_help) {
        help(_argc, _argv);
        // No context initializations
        pipeline(_argc, _argv, Context.error, Action.init);
        // no context finalization
        pipeline(_argc, _argv, Context.error, Action.exit);
        return 0;
    }

    // Error pipeline
    if (_error) {
        if (_error == 1) { // Environment variable missing
            if ((sd.erun.is_master = lock_master(path_tmp, sd.erun.job_id))) {
                plug_verbose(_sp, 0, "missing environment vars, is the EAR module loaded? (only master prints this message)");
                // Returning -1 is serious, so we announce that in all processes.
                sleep(2);
                // But only the real master can do the cleaning
                if (sd.erun.is_master == 1) {
                    folder_clean(sd.erun.job_id);
                }
            }
            return 0;
        }
        if (_error == 2) { // Going inactive because detected SRUN
            if ((sd.erun.is_master = lock_master(path_tmp, sd.erun.job_id)) == 1) {
                plug_verbose(_sp, 0, "detected SRUN, going inactive (only master prints this message)");
            } else if (sd.erun.is_master == -1) {
                plug_verbose(_sp, 0, "detected SRUN, but there is an error %s", state_msg);
            }
            // All processes of type master/slave/error launch the program
            execute(_argc, _argv);
            // But only the real master can do the cleaning
            if (sd.erun.is_master == 1) {
                folder_clean(sd.erun.job_id);
            }
            return 0;
        }
    } // _error

    // Is master?
    switch ((sd.erun.is_master = lock_master(path_tmp, sd.erun.job_id))) {
        case 1: // Case master
            plug_verbose(_sp, 2, "subject '%d' is erun master? '%d' (got the lock file)", getpid(), sd.erun.is_master);
            // Read old step id in master.step.id
            sd.erun.step_id = master_getstep(sd.erun.job_id, sd.erun.step_id);
            break;
        case 0: // Case not master
            plug_verbose(_sp, 3, "subject '%d' is erun master? '%d' (missed the lock file and then spinlock)", getpid(), sd.erun.is_master);
            // Spinlock over lock.slave
            spinlock_slave(sd.erun.job_id);
            // Get slave.step.id
            sd.erun.step_id = slave_getstep(sd.erun.job_id, sd.erun.step_id);
            break;
        case -1: // Case some error with lock files
            plug_error(_sp, "%s", state_msg);
            execute(_argc, _argv);
            return 0;
    }

    // Creating step
    step(_argc, _argv);
    // Local context initialization
    pipeline(_argc, _argv, Context.srun, Action.init);
    // Remote context initialization
    pipeline(_argc, _argv, Context.remote, Action.init);

    if (sd.erun.is_master) {
        // Free lock.step
        unlock_slave(sd.erun.job_id, sd.erun.step_id);
    }

    execute(_argc, _argv);

    if (sd.erun.is_master) {
        sleep(1);
    }

    // Remote context finalization
    pipeline(_argc, _argv, Context.remote, Action.exit);
    // Local context finalization
    pipeline(_argc, _argv, Context.srun, Action.exit);

    if (sd.erun.is_master) {
        files_clean(sd.erun.job_id, sd.erun.step_id);
    }

    return 0;
}
